### PR TITLE
Enable sharing meta tensors between processes

### LINKED ISSUE #129436: Enable multiprocess sharing for `nn.Module`s that have meta tensors
### 🚀 The feature, motivation and pitch

In TorchRL, there are a couple `nn.Module`s that contain a set of other submodules, where one of the submodules has `device='meta'`, and the rest are on a real device. I don't exactly know the details, but it looks like TorchRL uses the meta submodule to facilitate vmapping across the submodules that are on real devices.

The problem is that since PyTorch doesn't allow meta tensors to be shared between processes, an error is raised if the user tries to share them between processes, as shown in https://github.com/pytorch/rl/issues/2180.

I suppose there might be at least two slightly different possible solutions:

1. Enable `Tensor(..., device='meta')` to be shared
2. Just enable sharing `nn.Module`s that contain meta tensors, by avoiding trying to share a tensor if it is on the meta device

### Alternatives

_No response_

### Additional context

_No response_

cc @VitalyFedyunin @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @ezyang @eellison @bdhirsh @vmoens

### PR DESCRIPTION
**Pull request by kurtamohler**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129520

Fixes #129436

cc @VitalyFedyunin @ezyang @eellison @bdhirsh @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1894,8 +1894,10 @@ def _clone_input(value):
             example_value = wrap_to_fake_tensor_and_record(
                 example_value, tx=tx, **kwargs
             )
-        if isinstance(example_value, torch.Tensor) and (
-            maybe_get_fake_mode(example_value) is not tx.fake_mode
+        if (
+            isinstance(example_value, torch.Tensor)
+            and example_value.device.type != "meta"
```
Comment (kurtamohler) [inline torch/_dynamo/variables/builder.py:1899]: I'm not completely sure if this is the right thing to do here, but it makes the dynamo tests pass

**Comment by ezyang** [comment]
holler if you need me to look

**Comment by kurtamohler** [comment]
> holler if you need me to look

Sure, that would be good @ezyang. But if you're still on PTO, I think it can wait

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**